### PR TITLE
#26 Update Triggers & Actions to use evaluateExpression function

### PR DESCRIPTION
**Pull request by CarlosNZ**
Note: in order to test, you'll need to re-initialise the database with the modified `insertData.js` in here, as there were malformed JSON strings in the last version.

To test: Run server, then in an SQL interface (e.g. pgAdmin) run this query:
```
UPDATE application SET "trigger" = 'onApplicationSubmit' WHERE template_id = 2
```
You should see a message written to the console from the `consoleLog` plugin.

### Tests:
- [ ] Applying trigger to `application` table causes `consoleLog` Action to run (as above)
- [ ] Action evaluator respects the `condition` parameter (In the table `template_action`, change the `condition` for the only record to `{"value": false}` and confirm that the Action doesn't run when triggered.

Change the JSON value in `condition` to more complex expressions and confirm the Action is triggered correctly (i.e. the `evaluateExpression` function is working correctly):
- [ ] `condition = {"operator":"=","children":[{"value":6},{"value":6}]}` (This should be `true`, so Action will run)
- [ ] `condition = {"operator":"REGEX", "children":[{"value":"carl@sussol.net"},{"value":"^[A-Za-z0-9.]+@[A-Za-z0-9]+\\.[A-Za-z0-9.]+$"}]}`(This should be `true`)
- [ ] `condition = {"operator":"=","children":[{"value":13},{"operator":"+","children":[{"value":6},{"value":6}]}]}` (`false`, so Action should not run)

### CODE REVIEW

**Review by nmadruga — COMMENTED** [review]
Just one comment, mostly to clarify what is done in that function. All good with tests.

## Tests:

### 1
#### Ran the insertData and the server. 
```
yarn database init
yarn dev_pg
```
#### Then triggered using pgAdmin:
`UPDATE application SET "trigger" = 'onApplicationSubmit' WHERE template_id = 2`
- [x] Applying trigger to `application` table causes `consoleLog` Action to run (as above)

### 2
#### Changed the existing template_action to have condition as **false**.
`UPDATE template_action SET "condition" = '{"value": false}' WHERE id = 1`
#### Then triggered using pgAdmin:
`UPDATE application SET "trigger" = 'onApplicationSubmit' WHERE template_id = 2`
- [x] Action evaluator respects the `condition` parameter (In the table `template_action`, change the `condition` for the only record to `{"value": false}` and confirm that the Action doesn't run when triggered.

### 3
#### Changed the existing template_action to have condition as described (to be evaluated to **true**)
`UPDATE template_action SET "condition" = '{"operator":"=","children":[{"value":6},{"value":6}]}' WHERE id = 1`
#### Then triggered using pgAdmin:
`UPDATE application SET "trigger" = 'onApplicationSubmit' WHERE template_id = 2`
- [x] Change the JSON value in `condition` to more complex expressions and confirm the Action is triggered correctly (i.e. the `evaluateExpression` function is working correctly): `condition = {"operator":"=","children":[{"value":6},{"value":6}]}` (This should be `true`, so Action will run)

### 4
#### Changed the existing template_action to have condition as described (to be evaluated to **true**)
`UPDATE template_action SET "condition" = '{"operator":"REGEX", "children":[{"value":"carl@sussol.net"},{"value":"^[A-Za-z0-9.]+@[A-Za-z0-9]+\\.[A-Za-z0-9.]+$"}]}' WHERE id = 1`
#### Then triggered using pgAdmin:
`UPDATE application SET "trigger" = 'onApplicationSubmit' WHERE template_id = 2`
- [x] Change the JSON value in `condition` to more complex expressions and confirm the Action is triggered correctly (i.e. the `evaluateExpression` function is working correctly):`condition = {"operator":"REGEX", "children":[{"value":"carl@sussol.net"},{"value":"^[A-Za-z0-9.]+@[A-Za-z0-9]+\\.[A-Za-z0-9.]+$"}]}`(This should be `true`)

### 5
#### Changed the existing template_action to have condition as described (to be evaluated to **false**)
`UPDATE template_action SET "condition" = '{"operator":"=","children":[{"value":13},{"operator":"+","children":[{"value":6},{"value":6}]}]}' WHERE id = 1`
#### Then triggered using pgAdmin:
`UPDATE application SET "trigger" = 'onApplicationSubmit' WHERE template_id = 2`
- [x] Change the JSON value in `condition` to more complex expressions and confirm the Action is triggered correctly (i.e. the `evaluateExpression` function is working correctly):`condition = {"operator":"=","children":[{"value":13},{"operator":"+","children":[{"value":6},{"value":6}]}]}` (`false`, so Action should not run)

**Comment by nmadruga** [comment]
Please also have a look on the conflict please @CarlosNZ. You can merge locally the master with your branch to resolve it.

File: src/components/triggersAndActions.ts [hunk 1]
```diff
@@ -87,15 +88,20 @@ export async function processTrigger(client: Client, payload: TriggerPayload) {
...
-    actions.forEach((action: Action) => {
+    for (const action of actions) {
       const parameter_queries = action.parameter_queries
-      Object.keys(parameter_queries).forEach((key) => {
-        parameter_queries[key] = evaluateExpression(parameter_queries[key])
-      })
-      // console.log(parameter_queries);
-    })
+      for (const key in parameter_queries) {
+        parameter_queries[key] = await evaluateExpression(parameter_queries[key])
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/triggersAndActions.ts:101]:
```
   // Evaluate parameters for each Action
    for (const action of actions) {
      const parameter_queries = action.parameter_queries
      for (const key in parameter_queries) {
        parameter_queries[key] = await evaluateExpression(parameter_queries[key])
      }
    }
```
The Actions loaded from the template have expressions for their `parameters`. This loop just evaluates each expression and replaces the value in each `action.parameter_queries` with its evaluated output. The `parameter_queries` variable is not used later as it's just used in this loop to update the `action.parameter_queries` values in-place.

**Comment by CarlosNZ** [comment]
> Please also have a look on the conflict please @CarlosNZ. You can merge locally the master with your branch to resolve it.

Done :)

File: src/components/triggersAndActions.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/triggersAndActions.ts:101]: Ah! so it is actually updating what is inside the `action.parameter_queries`? I was thinking that `parameter_queries` was just a copy, but from what you are saying it seems that you get a const of the object, but the values it is pointing to are reference. Is that so? The code doesn't make this so clear.

[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/triggersAndActions.ts]: Yeah, I guess it's not necessary to make a new variable.

Have now amended it to:

```
// Evaluate parameters for each Action
    for (const action of actions) {
      for (const key in action.parameter_queries) {
        action.parameter_queries[key] = await evaluateExpression(action.parameter_queries[key])
      }
    }
```

**Pull request merged**